### PR TITLE
feat(learning): Integrate GLTorch with GraphScope in Server-Client Mode with K8S Deployment Supports

### PR DESCRIPTION
**Pull request by husimplicity**
## What do these changes do?

As titled.

### CODE REVIEW

File: python/graphscope/learning/gl_torch_examples/k8s/k8s_launch.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    },
+    node_labels={
+        "paper": "label",
+    },
+    edge_dir="out",
+    random_node_split={
+        "num_val": 0.1,
+        "num_test": 0.1,
+    },
+    master_id=0,
```
Comment (LiSu) [inline python/graphscope/learning/gl_torch_examples/k8s/k8s_launch.py:56]: Remove master_id

File: coordinator/gscoordinator/cluster_builder.py [hunk 2]
```diff
@@ -413,10 +458,14 @@ def get_learning_service(self, object_id, start_port):
         )
         return service
 
-    def get_learning_ports(self, start_port):
+    def get_graphlearn_ports(self, start_port):
         num_workers = self._num_workers
         return [i for i in range(start_port, start_port + num_workers)]
 
+    def get_graphlearn_torch_ports(self, start_port):
+        num_workers = 4
```
Comment (LiSu) [inline coordinator/gscoordinator/cluster_builder.py:466]: Properly name this parameter

File: coordinator/gscoordinator/cluster_builder.py [hunk 3]
```diff
@@ -400,7 +429,23 @@ def get_engine_headless_service(self):
     def get_learning_service(self, object_id, start_port):
```
Comment (LiSu) [inline coordinator/gscoordinator/cluster_builder.py:429]: This function is used with a new name "get_graphlearn_service" in kubernetes_launcher.py line 878

File: python/graphscope/learning/gl_torch_examples/local.py
[the remark this answers is not in the document]
Comment (husimplicity) [reply python/graphscope/learning/gl_torch_examples/local.py:1]: recovered

File: python/graphscope/learning/gl_torch_examples/k8s/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+### 1. Prepare data and code
+
+We use [Kubeflow](https://github.com/kubeflow/training-operator) to deploy **GLT** jobs on K8s clusters. Make sure Kubeflow is installed before running the following examples.
+
+### 2. Launch a GraphScope cluster on k8s and start training.
+`client.yaml` is an example of launching GLTorch jobs with a default setting, 2 servers and 2 workers, using pytorch-operator.
+`client.py` is the GLT training script example, and `k8s_launch.py` is the script to launch the GraphScope k8s cluster and the client jobs.
+
+First, please specify the corresponding file paths in the `gs.graphlearn_torch`:
```
[the remark this answers is not in the document]
Comment (husimplicity) [reply python/graphscope/learning/gl_torch_examples/k8s/README.md]: clarified

File: coordinator/gscoordinator/launch_graphlearn_torch.py [hunk 5]
```diff
@@ -49,36 +52,43 @@ def run_server_proc(proc_rank, handle, config, server_rank, dataset):
...
-    dataset = glt.distributed.DistDataset(edge_dir=edge_dir)
+
+    dataset = glt.distributed.DistDataset(
+        edge_dir=edge_dir,
+        num_partitions=handle["num_servers"],
+        partition_idx=server_rank,
+        node_pb=glt.data.VineyardPartitionBook(
+            str(handle["vineyard_socket"]),
+            str(handle["fragments"][server_rank]),
+            "paper",
```
[the remark this answers is not in the document]
Comment (husimplicity) [reply coordinator/gscoordinator/launch_graphlearn_torch.py]: done

File: python/graphscope/learning/gl_torch_examples/k8s/k8s_launch.py (hunk 1 above)
Comment (husimplicity) [reply python/graphscope/learning/gl_torch_examples/k8s/k8s_launch.py]: removed

File: coordinator/gscoordinator/cluster_builder.py (hunk 2 above)
Comment (husimplicity) [reply coordinator/gscoordinator/cluster_builder.py]: renamed

File: coordinator/gscoordinator/cluster_builder.py (hunk 3 above)
Comment (husimplicity) [reply coordinator/gscoordinator/cluster_builder.py]: fixed

**Review by LiSu — DISMISSED** [review]

File: python/graphscope/deploy/kubernetes/cluster.py [hunk 6]
```diff
@@ -183,8 +183,8 @@ def _create_role_and_binding(self):
             role = ResourceBuilder.get_role(
                 name=self._role_name,
                 namespace=self._namespace,
-                api_groups=",apps,extensions",  # The leading comma is necessary, represents for core api group.
-                resources="configmaps,deployments,deployments/status,statefulsets,statefulsets/status,endpoints,events,pods,pods/log,pods/exec,pods/status,services,replicasets",  # noqa: E501
+                api_groups=",apps,extensions,kubeflow.org",  # The leading comma is necessary, represents for core api group.
+                resources="configmaps,deployments,deployments/status,statefulsets,statefulsets/status,endpoints,events,pods,pods/log,pods/exec,pods/status,services,replicasets,pytorchjobs",  # noqa: E501
```
Comment (siyuan0322) [inline python/graphscope/deploy/kubernetes/cluster.py:187]: Will this work normally in a cluster without kubeflow installed.

**Review by siyuan0322 — APPROVED** [review]

**Review by LiSu — DISMISSED** [review]

File: coordinator/gscoordinator/kubernetes_launcher.py [hunk 7]
```diff
@@ -1356,13 +1556,75 @@ def close_learning_instance(self, object_id):
...
+                name="graphlearn-torch-client-config",
+                namespace=self._namespace,
+            )
+            print(
+                f"ConfigMap graphlearn-torch-client-config deleted. Response: {response}"
+            )
+        except K8SApiException as e:
+            print(
+                f"Exception when calling CoreV1Api->delete_namespaced_config_map: {e}"
+            )
```
Comment (sighingnow) [inline coordinator/gscoordinator/kubernetes_launcher.py:1627]: Use `logger.info` for logging.

File: coordinator/gscoordinator/kubernetes_launcher.py [hunk 8]
```diff
@@ -1318,30 +1365,183 @@ def _distribute_learning_process(
...
+                version=version,
+                namespace=namespace,
+                plural=plural,
+                body=pytorch_job_manifest,
+            )
+            print(api_response)
+        except K8SApiException as e:
+            print(
+                f"Exception when calling CustomObjectsApi->create_namespaced_custom_object: {e}"
+            )
```
Comment (sighingnow) [inline coordinator/gscoordinator/kubernetes_launcher.py:1472]: I suggest you `reraise` here and don't continue silently.

If creating the `PytorchJob` failed, the following watch loop will never return.

File: coordinator/gscoordinator/kubernetes_launcher.py [hunk 9]
```diff
@@ -209,10 +219,15 @@ def __init__(self, config: Config):
...
+        self._graphlearn_start_port = 8000
+        # 9001 ~ 10001 is exposed
+        self._graphlearn_torch_start_port = 9001
 
         self._graphlearn_services = {}
-        self._learning_instance_processes = {}
+        self._graphlearn_instance_processes = {}
+
+        self._graphlearn_torch_services = {}
+        self._graphlearn_torch_instance_processes = {}
```
Comment (sighingnow) [inline coordinator/gscoordinator/kubernetes_launcher.py:231]: Can graphlearn and graphlearn_torch share the same set of port/processes/services to make these part simpler?

**Review by sighingnow — DISMISSED** [review]
LGTM, with some minor comments.

File: coordinator/gscoordinator/kubernetes_launcher.py (hunk 8 above)
Comment (husimplicity) [reply coordinator/gscoordinator/kubernetes_launcher.py:1472]: fixed

File: coordinator/gscoordinator/kubernetes_launcher.py (hunk 9 above)
Comment (husimplicity) [reply coordinator/gscoordinator/kubernetes_launcher.py:231]: We will deprecate graphlearn later, so the separated implementation for now is fine :)

File: python/graphscope/deploy/kubernetes/cluster.py (hunk 6 above)
Comment (husimplicity) [reply python/graphscope/deploy/kubernetes/cluster.py:187]: seems normal as tested

**Review by sighingnow — APPROVED** [review]

**Pull request merged**